documentation/src/main/asciidoc/reference/introduction.adoc

```diff
@@ -401,6 +401,41 @@ values!
 When you use Hibernate Reactive in Quarkus, these requirements are relaxed,
 and you can use public fields instead of getters and setters if you prefer.
 
+=== `equals()` and `hashCode()`
+
+Entity classes should override `equals()` and `hashCode()`. People new to
+Hibernate or JPA are often confused by exactly which fields should be
+included in the `hashCode()`, so please keep the following principles in
+mind:
+
+- You should not include mutable fields in the hashcode, since that would
+  require rehashing any collection containing the entity whenever the field
+  is mutated.
+- It's not completely wrong to include a generated identifier (surrogate
+  key) in the hashcode, but since the identifier is not generated until
+  the entity instance is made persistent, you must take great care to not
+  add it to any hashed collection before the identifier is generated. We
+  therefore advise against including any database-generated field in the
+  hashcode.
+
+It's OK to include any immutable, non-generated field in the hashcode.
+
+TIP: We therefore recommend identifying a _natural key_ for each entity,
+that is, a combination of fields that uniquely identifies an instance of
+the entity, from the perspective of the data model of the program. The
+business key should correspond to a unique constraint on the database,
+and to the fields which are included in `equals()` and `hashCode()`.
+
+IMPORTANT: If you can't identify a natural key, it's _highly likely_
+that you have not thought hard enough about your data model, or that
+there's something actually wrong with your model. If your entity doesn't
```

Comment:
Maybe it's just me, but I don't feel comfortable telling a user that they haven't thought hard enough about an issue.

Reply:
We could keep it short and say that without a natural identifier, it's not possible to distinguish between two objects with the same fields and that usually, they could add a column containing a generated identifier.

Reply:
But that's not what I'm trying to say here: I've had too many arguments with people who insist that their table has no natural key, and so they want to use the surrogate key as the hashcode, which causes problems. 

What I'm trying to get them to do is think harder about their model and identify the natural key.

And yeah, I have never ever been shy about telling users when they're wrong.

Reply:
I don't see how telling somebody "Think harder" is going to help them.
Has it ever worked in the past?

Anyway, I think you can rewrite the same paragraph without mentioning the user:
```
IMPORTANT: If your model doesn't have a natural key, it's _highly likely_ that there's something wrong with it.
```
> And yeah, I have never ever been shy about telling users when they're wrong.

Really, I haven't noticed :-)
That said, no user has made any statement, so I think we can give them the benefit of the doubt.  This is not twitter after all.

Reply:
> Has it ever worked in the past?

Absolutely yes!

Reply:
Really, I think it's much more important to give people good advice that will help them have success with Hibernate than to avoid the potential for momentary slight emotional discomfort.

Reply:
I don't see how the two things are related. It's possible to give good advice without being patronizing. 

> > Has it ever worked in the past?
> 
> Absolutely yes!

I doubt it.

I think it's something we will never agree on.
Feel free to merge whichever version you prefer.